credentials/tls/certprovider/meshca/plugin_test.go

```diff
@@ -0,0 +1,335 @@
+		}
+		return grpc.Dial(uri, grpc.WithInsecure())
+	}
+	defer func() { grpcDialFunc = origDialFunc }()
+
+	// Override the plugin's backoff function with this, so that we can verify
+	// that a backoff actually was triggered.
+	origBackoffFunc := backoffFunc
+	boCh := testutils.NewChannelWithSize(maxErrCount)
+	backoffFunc = func(v int) time.Duration {
+		boCh.Send(v)
+		return 0
+	}
+	defer func() { backoffFunc = origBackoffFunc }()
+
+	// Set the MeshCA targetURI in the plugin configuration to point to our fake
+	// MeshCA.
+	cfg := proto.Clone(goodConfigFullySpecified).(*configpb.GoogleMeshCaConfig)
+	cfg.Server.GrpcServices[0].GetGoogleGrpc().TargetUri = addr
+	inputConfig := makeJSONConfig(t, cfg)
+	prov, err := certprovider.GetProvider(pluginName, inputConfig, certprovider.Options{})
+	if err != nil {
+		t.Fatalf("certprovider.GetProvider(%s, %s) failed: %v", pluginName, cfg, err)
+	}
+	defer prov.Close()
+
+	// We don't really care about the exact key material returned here. All we
+	// care about is whether we get any key material at all, and that we don't
+	// get any errors.
+	ctx, cancel := context.WithTimeout(context.Background(), 3*defaultTestTimeout)
+	defer cancel()
+	if _, err = prov.KeyMaterial(ctx); err != nil {
+		t.Fatalf("provider.KeyMaterial(ctx) failed: %v", err)
+	}
+
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	for i := 0; i < maxErrCount; i++ {
+		if _, err := boCh.Receive(ctx); err != nil {
+			t.Fatalf("plugin failed to backoff after error from fake server: %v", err)
+		}
+	}
+}
+
+func (s) TestCreateCertificateWithRefresh(t *testing.T) {
+	addr, cancel := setupFakeMeshCA(t, false, true)
+	defer cancel()
+
+	// Override the dial func to make sure that the plugin is dialing the
+	// expected fakeCA, and also to use insecure creds here.
+	origDialFunc := grpcDialFunc
+	grpcDialFunc = func(uri string, _ ...grpc.DialOption) (*grpc.ClientConn, error) {
+		if uri != addr {
+			t.Fatalf("plugin dialing MeshCA at %s, want %s", uri, addr)
+		}
+		return grpc.Dial(uri, grpc.WithInsecure())
+	}
+	defer func() { grpcDialFunc = origDialFunc }()
+
+	// Set the MeshCA targetURI in the plugin configuration to point to our fake
+	// MeshCA.
+	cfg := proto.Clone(goodConfigFullySpecified).(*configpb.GoogleMeshCaConfig)
+	cfg.Server.GrpcServices[0].GetGoogleGrpc().TargetUri = addr
+	inputConfig := makeJSONConfig(t, cfg)
+	prov, err := certprovider.GetProvider(pluginName, inputConfig, certprovider.Options{})
+	if err != nil {
+		t.Fatalf("certprovider.GetProvider(%s, %s) failed: %v", pluginName, cfg, err)
+	}
+	defer prov.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 3*defaultTestTimeout)
+	defer cancel()
+	km1, err := prov.KeyMaterial(ctx)
+	if err != nil {
+		t.Fatalf("provider.KeyMaterial(ctx) failed: %v", err)
+	}
+
+	// Sleep for a little while to make sure that the certificate refresh
+	// happens in the background.
+	time.Sleep(2 * shortTestCertLife)
```

Comment:
Can this be avoided by blocking on a channel until they are refreshed?  If you can't tell, I hate wasted time in tests; any time-related things should be removed if at all possible, including things like expiration *times*, and use something event-based instead.  There are times when this is more trouble to avoid than it's worth to have hooks in the necessary places, but if it can be avoided, please try!